styleguide/angular.md

```diff
@@ -0,0 +1,178 @@
+# Joinbox Angular Guidelines
+This document describes general guidelines for creating Angular code at Joinbox. The explanations are mainly important for legacy Angular code (_i.e._ 1.6.x versions as used for the SBC project).
+
+## Angular 1.6.x
+The following sections describe on how to handle, refactor and create Angular code in the future. It will not explain core concepts. Instead, the following sections explain how and why to move forward with the legacy codebase.
+
+> **Note:** Upgrading the Angular version on a such extensive codebase as the SBC project resides between being impossible and absolutely impossible. So what we are looking for is a way forward that allows us to leverage the old code to be able to benefit from the concepts that make newer Angular versions _better_.  
+
+### Rules of Thumb
+
+Let's start with an overview (please see the other sections if you're not aware of the concepts):
+
+1. If your project allows it, use `ES6` (precompile your code if necessary).
+2. Name your files consistently!
+3. Bundle logic (especially shared logic) in services!
+4. Write components (not directives)!
+5. Use bindings wisely especially in presence of deep nesting and loops!
+6. Use `ng-if` as guard for view rendering and to control its timing!
+
+### Components
+In general, Angular supports  a variety of [concepts](https://docs.angularjs.org/guide/concepts). The most important one for our legacy code are [components](https://docs.angularjs.org/guide/component). They are the most basic common denominator of Angular 1.x.x and >2.0.0.
+
+Components provide a way cleaner interface than directives did in earlier versions of Angular!
+
+> **Rule of thumb:** Don't write directives! Refactor your directives towards a component. As soon as one has done it once, it will be straight forward (..._most of the time_).   
+
+> **To Clarify:** ~~Where do we manipulate the DOM in components?~~ DOM manipulations happen during the `$postLink` lifecycle hook.  
+
+1. Forget about scope inheritance! It is evil. All directives have an isolated scope! All the data that is passed into the component is clearly defined in the `bindings` section of the components definition.
+2. Create services to share logic and load data if necessary.
+
+### Structure
+Reworking a codebase towards components brings up an old discussion about how to structure Angular code. Some prefer to group the code together based on their type of content (_e.g._ having a `service` folder, a `directive` folder ...). 
+
+Even though this might be subjective I'd go for a functional/domain based structure:
```

Comment:
Yeah the list of folders is not final. It's just an example.

> What benefit do you see from grouping the code files domain based?

It makes it easier for me to find code. I do not really care how something is implemented (if it is a directive, a component, whatever), i want to know what it does.

Most of the time we create components in a certain domain. _E.g._ we are creating services in the Typo3 project for all entities in the services folder instead of the folder where the rest concerning the corresponding entity is stored (it is even worse in the admin).